libevmasm/PeepholeOptimiser.cpp

```diff
@@ -46,6 +46,14 @@ struct ApplyRule
 {
 };
```

Comment:
```suggestion
{
	template <size_t... Indices>
	static bool applyRule(AssemblyItems::const_iterator _in, std::back_insert_iterator<AssemblyItems> _out, std::index_sequence<Indices...>)
	{
		return Method::applySimple(_in[Indices]..., _out);
	}

	static bool applyRule(AssemblyItems::const_iterator _in, std::back_insert_iterator<AssemblyItems> _out)
	{
		return applyRule(_in, _out, make_index_sequence<Arguments>{});
	}
};
```
and removing all the explicit specializations should just work, shouldn't it?

Reply:
maybe it does in the meantime, let me check.

Reply:
We could probably simplify it even further, s.t. there is no need for spelling out the size anywhere including in ``SimplePeepholeOptimizerMethod<..., WindowSize>``... but not sure that's worth it...

Reply:
```c++
template<typename Method>
struct GetWindowSize;
template<typename R, typename... Args>
struct GetWindowSize<R(Args...)>
{
	static constexpr auto value = sizeof...(Args) - 1;
};

template <class Method>
struct SimplePeepholeOptimizerMethod
{
	static bool apply(OptimiserState& _state)
	{
		static constexpr size_t WindowSize = GetWindowSize<decltype(Method::applySimple)>::value;
		if (
			_state.i + WindowSize <= _state.items.size() &&
			ApplyRule<Method, WindowSize>::applyRule(_state.items.begin() + static_cast<ptrdiff_t>(_state.i), _state.out)
		)
		{
			_state.i += WindowSize;
			return true;
		}
		else
			return false;
	}
};
```

Reply:
With that there shouldn't be any need for ``ApplyRule`` at all anymore and the window size template argument can be dropped from all methods...